nixos/modules/system/activation/switch-to-configuration.pl

```diff
@@ -933,16 +934,51 @@ sub filter_units {
 }
 
-# Start all active targets, as well as changed units we stopped above.
-# The latter is necessary because some may not be dependencies of the
-# targets (i.e., they were manually started).  FIXME: detect units
+# Start all active _targets_, as well as changed _units_ we stopped above.
+# Both are included in `units_to_start`.
+# Starting the changed units explicitly is necessary because
+# some may not be dependencies of the
+# targets (i.e., they were manually started).
+#
+# Note [systemd-timers-cancelling-stop-commands]:
+# We need to start the changed units with `restart` instead of `start`.
+# This is because of systemd timers, which can make our life miserable:
+#     https://github.com/NixOS/nixpkgs/issues/49415
+# If during `systemctl stop` above, a timer restarts a unit (directly or
+# as a dependency), systemctl will cancel the stop and execute the restart.
+# The stop command in that case prints:
+#     Job for myjob.service canceled
+# If a timer-caused restart cancels a stop, the unit will have been
+# restarted with the unit definition from before the configuration switch,
+# so it will be outdated (e.g. run the wrong binary or config).
+# If we only `start`ed such a unit here, it would continue running outdated.
+# Thus, we must `restart` the units we stopped above.
+# But we need to do that only with the units we stopped above,
+# not all `units_to_start`: If we `restart`ed all `.target`s,
+# this would do too much, e.g. kill the network.
+# A minor drawback of this approach is that units that fail the `restart`
+# get a second chance with the subsequent `start`.
+# This may be confusing to users when reading logs or designing timeouts.
+# Unfortunately, we must do it this way, because otherwise a fundamental promise
+# of NixOS is broken (congruent configuration management:
+# "the system runs the declared services, not some old leftover ones").
+# A better solution would be if systemd provided a command where one
+# could, within a single systemd transaction, say which units should
+# be started and which ones restarted.
+# For what "transaction" means, see:
+#     https://github.com/systemd/systemd/issues/10626
+#
+# FIXME: detect units
 # that are symlinks to other units.  We shouldn't start both at the
 # same time because we'll get a "Failed to add path to set" error from
 # systemd.
 my @units_to_start_filtered = filter_units(\%units_to_start);
 if (scalar(@units_to_start_filtered)) {
-    print STDERR "starting the following units: ", join(", ", @units_to_start_filtered), "\n"
+    # See note [systemd-timers-cancelling-stop-commands].
+    print STDERR "starting (via restart, see #49415) the following units: ", join(", ", sort(@units_to_start_filtered)), "\n";
+    system("$new_systemd/bin/systemctl", "restart", "--", sort(@units_to_start_filtered)) == 0 or $res = 4;
 }
+print STDERR "ensuring the following units and targets are started: ", join(", ", sort(keys(%units_to_start))), "\n";
```

Comment:
Is this not potentially a quite long list. I thought this is what the filter above was for?

Reply:
It's a medium long list on my servers, e.g. 3-4 lines in my 100-char terminal.

I believe it is worth showing, because:

* It's useful for debugging what's going on. The `systemctl start` done here is one of the 3 key actions done in configuration switching (the other ones being `stop` and `restart`), and depending on what your units/targets do, it may be the last line you see before your server loses the network or desktop loses the GPU output. In those cases, it can be extremely helpful to see what's being started.
* This not being an invisble `start` invocation will hopefully reduce the number of issues people can't diagnose/solve by themselves, and thus number of issues they file. Most NixOS users understand `systemctl` and can reproduce issues when being told what `systemctl` commands to run, but much fewer users are able/willing to read/modify Perl to be able see the invocation.

I added it as a separate log line in addition to the previous one that lists only the stopped units, so that the short, most important line stays easy to read for the user. Then the longer line follows, which the user can ignore if their system didn't go down.